File: Secure Code Review/SSRF/Java/ssrf-fix.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.InetAddress;
import java.util.regex.Pattern;
import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

@WebServlet("/fetchData")
public class SecureSSRFServlet extends HttpServlet {

    // Allowlist of trusted domains
    private static final String[] ALLOWED_DOMAINS = {"example.com", "trusted.com"};

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        String targetUrl = request.getParameter("url"); // User input
        
        if (targetUrl == null || targetUrl.isEmpty()) {
            response.getWriter().println("No URL provided!");
            return;
        }

        try {
            // Validate the URL
            URL url = new URL(targetUrl);
            if (!isUrlAllowed(url)) {
                response.getWriter().println("Invalid or disallowed URL.");
                return;
            }

            // Proceed with the request if URL is valid
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");

            BufferedReader in = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            StringBuilder responseContent = new StringBuilder();
            String inputLine;

            while ((inputLine = in.readLine()) != null) {
                responseContent.append(inputLine);
            }
            in.close();

            response.getWriter().println(responseContent.toString());
        } catch (Exception e) {
            response.getWriter().println("Error fetching URL: " + e.getMessage());
        }
    }

    // Validate the URL against allowed domains and block internal IPs
    private boolean isUrlAllowed(URL url) {
        try {
            // Validate against the allowlist
            String host = url.getHost();
            for (String allowedDomain : ALLOWED_DOMAINS) {
                if (host.endsWith(allowedDomain)) {
                    return true;
                }
            }

            // Prevent access to internal/private IP ranges
            InetAddress address = InetAddress.getByName(host);
            if (address.isSiteLocalAddress() || address.isLoopbackAddress()) {
                return false;
            }
        } catch (Exception e) {
            return false;
        }
        return false;
    }
}





















...........................................................................................................................................
///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

Explanation of Fixes Above snipped code

    Allowlist Validation:
        The isUrlAllowed method ensures only URLs pointing to trusted domains like example.com and trusted.com are allowed.
        This mitigates access to unauthorized or malicious domains.

    Block Internal IP Addresses:
        The InetAddress class is used to block requests to private or loopback IP ranges (e.g., 127.0.0.1 or 192.168.x.x).

    Input Validation:
        Check the structure and format of the URL before processing it.
        Avoid processing untrusted user input without verification.

Additional Recommendations

    Time-Out Settings:
        Limit the time for HTTP connections to prevent hanging requests.

    HTTP Methods:
        Restrict requests to safe methods like GET. Avoid exposing sensitive actions like POST.

    Content Security Policy (CSP):
        Enforce strict policies on where resources can be loaded from.

File: Secure Code Review/SSRF/.Net/ssrf-fix.cs
using System;
using System.IO;
using System.Net.Http;
using System.Net;
using Microsoft.AspNetCore.Mvc;

namespace SecureApp.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class FetchController : ControllerBase
    {
        private static readonly string[] AllowedDomains = { "example.com", "trusted.com" };

        [HttpGet]
        public IActionResult GetExternalData([FromQuery] string url)
        {
            if (string.IsNullOrEmpty(url))
            {
                return BadRequest("URL parameter is required.");
            }

            try
            {
                // Validate the URL
                if (!IsUrlAllowed(url))
                {
                    return BadRequest("Invalid or disallowed URL.");
                }

                // Use HttpClient to fetch data
                using (var httpClient = new HttpClient())
                {
                    var response = httpClient.GetStringAsync(url).Result;
                    return Ok(response);
                }
            }
            catch (Exception ex)
            {
                return StatusCode(500, $"Error fetching data: {ex.Message}");
            }
        }

        private bool IsUrlAllowed(string url)
        {
            try
            {
                var uri = new Uri(url);

                // Validate against allowed domains
                foreach (var domain in AllowedDomains)
                {
                    if (uri.Host.EndsWith(domain, StringComparison.OrdinalIgnoreCase))
                    {
                        return true;
                    }
                }

                // Prevent access to internal/private IPs
                var hostEntry = Dns.GetHostEntry(uri.Host);
                foreach (var ipAddress in hostEntry.AddressList)
                {
                    if (IPAddress.IsLoopback(ipAddress) || IsPrivateIP(ipAddress))
                    {
                        return false;
                    }
                }
            }
            catch
            {
                return false;
            }
            return false;
        }

        private bool IsPrivateIP(IPAddress ipAddress)
        {
            byte[] bytes = ipAddress.GetAddressBytes();
            return (bytes[0] == 10) || // 10.0.0.0 - 10.255.255.255
                   (bytes[0] == 172 && (bytes[1] >= 16 && bytes[1] <= 31)) || // 172.16.0.0 - 172.31.255.255
                   (bytes[0] == 192 && bytes[1] == 168); // 192.168.0.0 - 192.168.255.255
        }
    }
}















///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
...........................................................................................................................................
Explanation of how this Above snipped code Fixes

    Domain Allowlist:
        The AllowedDomains array ensures that only specific domains like example.com and trusted.com are allowed.

    Block Internal IPs:
        The IsPrivateIP function blocks requests to private IP ranges (e.g., 10.x.x.x, 192.168.x.x) and loopback addresses (127.x.x.x).

    Use HttpClient:
        A safe and modern HTTP client is used instead of the legacy WebRequest.

    Error Handling:
        Gracefully handles invalid or malicious URLs by returning an appropriate error response.

Additional Recommendations

    Timeouts:
        Set timeouts on HTTP requests to prevent hanging connections.

    Limit Response Size:
        Restrict the maximum size of responses to avoid excessive resource consumption.

    Content-Type Validation:
        Validate that the response's Content-Type matches expected types (e.g., application/json).

    Use Firewall Rules:
        Employ firewall rules to block access to sensitive internal resources.








